Fix Per-Monitor DPI awareness of some controls

MaartenBent opened the pull request:
Update the font of some buddy controls when the DPI changes. Fix the position of the statusbar after a DPI change. Add some changes that were suggested but left out from #1499.

Some sizes are cached to improve the speed of the library. These sizes become incorrect when the DPI changes. And are incorrect when a window is created on a display with a different DPI.
Fix this by checking if the current DPI is the same as the DPI that was used when calculating the size, otherwise recalculate the size.

I tried to keep it as fast as possible by caching `s_checkDPI` and only check it when `GetDpiForWindow` will be used (`GetDeviceCaps` is much slower). Please let me know if this is an acceptable approach.

MaartenBent committed:
Update wxDPIChangedEvent documentation

MaartenBent committed:
Require wxWindow parameter for wxNativeFontInfo constructor in MSW

Use the DPI of the window to determine the correct font pointSize.

To not break user code, add a default argument when not building the library.

MaartenBent committed:
Fix font of wxRadioBox after DPI change

MaartenBent committed:
Fix font and thumb size of wxSlider after DPI change

MaartenBent committed:
Fix wxStatusBar font and position after DPI changes

Update the font of the associated wxDC so ellipsization uses the correct font
size.

vadz commented on include/wx/fontutil.h:
```diff
@@ -118,11 +118,11 @@ class WXDLLIMPEXP_CORE wxNativeFontInfo
-    {
-    }
+    wxNativeFontInfo(const LOGFONT& lf_, const wxWindow* win
+#ifndef WXBUILDING
```
I think this is sufficiently non-trivial to merit an explanation, i.e. something like
```
// Preserve compatibility in the semi-public (i.e. private, but still
// unfortunately used by some existing code outside of the library) API
// by allowing to create wxNativeFontInfo from just LOGFONT, but ensure
// that we always specify the window, to use the correct DPI, when creating
// fonts inside the library itself.
```

vadz commented on include/wx/private/window.h:
```diff
@@ -33,6 +34,34 @@ inline wxSize GetAverageASCIILetterSize(const T& of_what)
 }
 
+inline bool HasDPIChanged(const wxWindowBase* const win, wxSize& usedDPI)
+{
```
I think we might need to do things differently depending on whether or not per-monitor DPI is supported elsewhere too, so I'd like to suggest adding some `extern bool wxHasPerWindowDPI()` that would be implemented in `src/msw/window.cpp` and writing this function as just
```cpp
inline bool HasDPIChanged(const wxWindowBase* const win, wxSize& usedDPI)
{
    if ( win && wxHasPerWindowDPI() )
    {
        const wxSize dpi = win->GetDPI();
        if ( dpi != usedDPI )
        {
            usedDPI = dpi;
            return true;
        }
    }

    return false;
}
```

MaartenBent replied:
A disadvantage of your example is that `wxHasPerWindowDPI` has to be implemented for each port. I can't use a base function and override it just for MSW. And adding another (private) function to `wxWindow` or `wxTopLevel` seems a bit unnecessary for just a one-time use.

Maybe add `wxHasPerWindowDPI` as a separate function in `wx/private/window.h`?

vadz replied:
Isn't this code MSW-specific anyhow? It's in `wx/msw/private.h` after all -- or can it be used by wxQt too here?

MaartenBent replied:
Its in the common `wx/private/window.h` because it has to be used from `common/wincmn.cpp` and `common/sizer.cpp`.
I'd rather add one `__WXMSW__` guard in this header than multiple in the cpp files.

vadz replied:
Sorry, I'm blind, I've somehow seen `msw` in the file name which is not there :-(

In any case, I'm fine with having `#ifdef __WXMSW__` here, I just thought it was unnecessary for some unfathomable reason.

I'd still use an early return instead of an extra bool variable, but this is just a minor style preference, of course (the code seems more clear like this to me).

vadz reviewed (commented):
Thanks!

As usual, I'm fine with applying this directly, but I can't prevent myself from thinking about 2 possible improvements:

First one is purely technical: could it be useful to have a `template <class T> DpiDependentValue` that would store the value of the specified type (`int`, `wxSize`, ...) and the DPI at which it was computed and provide `HasChanged()` method? The idea would be to replace 

```cpp
static wxSize s_defFontSize;
static wxSize s_usedDPI;
const bool dpiChanged = wxPrivate::HasDPIChanged(parent, s_usedDPI);

if ( s_defFontSize.x == 0 || dpiChanged )
    s_defFontSize = wxPrivate::GetAverageASCIILetterSize(*parent);
```

with more concise

```cpp
static DpiDependentValue<wxSize> s_defFontSize;
if ( s_defFontSize.HasDPIChanged(parent) )
    s_defFontSize.SetAtNewDPI(wxPrivate::GetAverageASCIILetterSize(*parent));
```
(this relies on updating the current DPI in `HasDPIChanged()` as a side effect, which results in the shortest possible code, but is a bit ugly, so maybe we should have `SetAtDPI()` taking `wxWindow*` from which the DPI is re-extracted as the first argument instead).

Second one is more high level: it's nice to fix functions such as `wxButton::GetDefaultSize()` to work at the DPI corresponding to the top-level app window, but it doesn't really help if this window is on another display, using a different DPI. So I think we need to either add `wxWindow*` arguments to all these functions, with the same hack for defaulting it to `NULL` only when _not_ building the library, or add new `GetDefaultSizeInWindow(wxWindow* win)` variants and deprecate the ones not taking the window. I'm not sure if it's really a good idea to do this, but OTOH I don't see how could using `GetDefaultSize()` without the associated window ever be correct in an application handling per-monitor DPI -- it's basically guaranteed to be wrong at least sometimes.

What do you think?

MaartenBent commented:
I like the short example using a template. I'll give it a try.

> such as `wxButton::GetDefaultSize()`

It might be the only one, I haven't found similar functions yet. Adding a `wxWindow` parameter seems like a good solution. The function is not used in wxWidgets sources, so I don't know if it is necessary to add the `WXBUILDING` hack.

vadz commented:
> I like the short example using a template. I'll give it a try.

AFAICS there shouldn't be any problems with implementing this, I was thinking along the lines of (completely untested...):
```cpp
template <typename T>
class DpiDependentValue
{
public:
    // Explicit initialization is needed if T is a primitive type.
    DpiDependentValue() : m_value() { }

    bool HasChanged(wxWindow* w)
    {
        // Ensure that we return true the first time we're called.
        if ( m_value != T() )
        {
            if ( !w->HasPerWindowDPI() )
                return false;

            const wxSize dpi = w->GetDPI();
            if ( dpi == m_dpi )
                return false;

            m_dpi = dpi;
        }

        return true;
    }

    // Note: value must not be the initial value, i.e. T().
    void SetAtNewDPI(T value)
    {
        m_value = value;
    }

private:
    wxSize m_dpi;
    T m_value;
};
```

> > such as `wxButton::GetDefaultSize()`
> 
> It might be the only one, I haven't found similar functions yet. Adding a `wxWindow` parameter seems like a good solution. The function is not used in wxWidgets sources, so I don't know if it is necessary to add the `WXBUILDING` hack.

I see at least the default sizer border. But maybe we could use some hack for passing border in DIPs to sizers? E.g. use the convention that negative values are in DIPs? This is definitely ugly, but so would be passing `window` to each and every `wxSizerFlags::GetDefaultBorder()` call...

MaartenBent committed:
Reset static sizes when DPI changes

MaartenBent committed:
Add optional wxWindow parameter to wxButtonBase::GetDefaultSize()

When per-monitor DPI is used, the default button size depends on the DPI of the
display. Use the window to determine this DPI.

MaartenBent commented:
> This is definitely ugly, but so would be passing `window` to each and every `wxSizerFlags::GetDefaultBorder()` call...

Other options could be adding `wxWindow` to `wxSizerFlags` constructor or adding `wxSizerFlags& Window(win){}` to associate it with a window.

But the controls I wanted to target in this PR should be fixed. If CI passes, it's ready for review and merge.

vadz commented:
The most reasonable choice would probably be to switch to using DIPs instead of pixels, but this risks creating so much confusion that I'm not sure how realistic is this in practice... Anyhow, I agree that this can be left for later, so let's merge these changes already.

Thanks a lot once again for your work on this!
